### PR TITLE
fix: pass search and hash to both url and as

### PR DESCRIPTION
**Pull request by chrisneven**
There's a bug where the the next router query object would get out of sync with the query params. In order to solve this you must pass `search` to the url in the router methods as well as to the `as` parameter. There's also something going wrong with hash values as they need to be passed as well.

Furthermore, I installed prettier (I hope you don't mind)

### CODE REVIEW

**Comment by amannn** [comment]
Hi @chrisneven and thank you very much for looking into this!

Could you adapt the example in the repo so we can verify this behaviour? There are unfortunately no tests in the repo yet, since the integration with Next.js would require E2E testing. But when the example handles this use case, we can verify it manually for the time being and possibly later introduce E2E testing based on the example.

> Furthermore, I installed prettier (I hope you don't mind)

Prettier is handled with the ESLint toolchain in this repo (see [the failed check](https://github.com/amannn/next-query-params/runs/4351004669?check_suite_focus=true)). If you have ESLint configured in your editor to fix issues automatically, it will behave the same as prettier. Furthermore you can autofix issues in the codebase with `yarn eslint src --fix`.

Can you please revert all the stylistic changes?

**Comment by chrisneven** [comment]
Hi @amannn, thank you for the feedback! 

I'll try to implement it later today

**Comment by chrisneven** [comment]
Hi @amannn, I've implemented your feedback. If there's anything else, let me know :)

File: packages/example/src/pages/index.tsx [hunk 1]
```diff
@@ -4,10 +4,12 @@ import {
   withDefault,
   BooleanParam
 } from 'next-query-params';
+import { useRouter } from 'next/dist/client/router';
```
Comment (amannn) [inline packages/example/src/pages/index.tsx:7]:
```suggestion
import { useRouter } from 'next/router';
```

Also please run ESLint with `--fix` in the example.

File: packages/example/src/pages/index.tsx [hunk 2]
```diff
@@ -34,6 +36,17 @@ export default function Index() {
           type="checkbox"
         />
       </label>
+
+      <hr/>
+      <p> after clicking on: <a href="#sync">a hash link</a> the url should contain the hash despite changing params</p>
```
Comment (amannn) [inline packages/example/src/pages/index.tsx:41]: Does this work for you? For me the hash is removed when query params change.

Kapture 2021-11-30 at 08 07 00

File: packages/example/src/pages/index.tsx [hunk 3]
```diff
@@ -34,6 +36,17 @@ export default function Index() {
...
+      <hr/>
+
+      <p>the router query object provided by next should contain all query params:</p>
+      <pre>
+        {JSON.stringify(router.query, null, 2)}
```
Comment (amannn) [inline packages/example/src/pages/index.tsx:46]: This doesn't seem to be changing for me.

Comment (chrisneven) [reply packages/example/src/pages/index.tsx]: Have you tried rebuilding the `next-query-params` package? 

```
yarn build
```

Comment (chrisneven) [reply packages/example/src/pages/index.tsx]: On the otherside you'll probably understand why I created this PR a little bit more :)

Comment (amannn) [reply packages/example/src/pages/index.tsx]: 🤦‍♂️ Of course, after rebuilding this works – thanks! I think I need a coffee …

Comment (chrisneven) [reply packages/example/src/pages/index.tsx]: haha same here, as I forgot the `eslint --fix`

File: packages/example/src/pages/index.tsx [hunk 4]
```diff
@@ -4,10 +4,12 @@ import {
   withDefault,
   BooleanParam
 } from 'next-query-params';
+import {useRouter} from 'next/dist/client/router';
```
Comment (amannn) [inline packages/example/src/pages/index.tsx:7]: Let's use the suggested way to import the router:

```suggestion
import {useRouter} from 'next/router';
```

File: packages/next-query-params/src/NextQueryParamProvider.tsx [hunk 5]
```diff
@@ -35,25 +35,23 @@ function NextQueryParamProvider({children, ...rest}: Props) {
...
-        ),
-      replace: ({search}: Location) => {
-        router.replace(
-          {pathname: router.pathname, query: router.query},
-          {search, pathname},
+  const history = useMemo(() => {
+    function createUpdater(routeFn: typeof router.push) {
+      return function updater({hash, search}: Location) {
+        routeFn(
+          {pathname: router.pathname, query: router.query, search, hash},
```
Comment (amannn) [inline packages/next-query-params/src/NextQueryParamProvider.tsx:42]: Adding `search` here is also a major improvement, thanks 👍

**Comment by amannn** [comment]
Released in `next-query-params@2.0.4`

**Pull request merged**